packages/suite/src/actions/suite/metadataActions.ts

```diff
@@ -311,68 +358,74 @@ const syncMetadataKeys = () => (dispatch: Dispatch, getState: GetState) => {
             payload: accountWithMetadata,
         });
     });
-    // todo: hm and what about devices??
+    // note that devices are intentionally omitted here - device receives metadata
+    // keys sooner when enabling labeling on device;
 };
 
 export const connectProvider = (type: MetadataProviderType) => async (dispatch: Dispatch) => {
-    try {
-        const provider = await getProvider({ type });
-        if (!provider) return;
-
+    let provider = await dispatch(getProvider());
+    if (!provider) {
+        provider = createProvider(type);
+    }
+    const isConnected = await provider.isConnected();
+    if (provider && !isConnected) {
         const connected = await provider.connect();
+        if (!connected) {
+            return;
+        }
+    }
 
-        if (connected) {
-            const credentials = await provider.getCredentials();
-            if (!credentials) {
-                return;
-            }
-            // TODO: toast errors
-            // set metadata reducer
-            dispatch({
-                type: METADATA.SET_PROVIDER,
-                payload: credentials,
-            });
+    const result = await provider.getCredentials();
+    if (!result.success) {
+        dispatch(handleProviderError(result, ProviderErrorAction.CONNECT));
+        return;
+    }
 
-            return true;
-        }
-    } catch (error) {
-        dispatch(handleProviderError(error));
+    dispatch({
+        type: METADATA.SET_PROVIDER,
+        payload: result.payload,
+    });
+
+    if (window.desktopApi) {
```

Comment:
move this to provider.connect + remove suite-native copy

Reply:
or electron layer

Reply:
I'll move it electron layer. It makes perfect sense there.